Add new filter json_encode to convert loglines to json-encoded raw

ekimekim opened the pull request:
It will also optionally try to parse the log line as JSON and include
that at the top level.

This is needed for our structured logging so we can emit json
to disk and have cernan forward it to kafka.

dparton commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,139 @@
+            }
+            metric::Event::Shutdown => {
+                res.push(metric::Event::Shutdown);
+            }
```
It looks to me like these last 3 match arms are all doing the same thing, so this could be simplified to:
```
event => { 
    res.push(event); 
}
```

ekimekim replied:
ah, right. I just copied this part from DelayFilter.

dparton commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,139 @@
+    /// Total number of logline processed
+    pub static ref JSON_ENCODE_LOG_PROCESSED: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
+    /// Total number of logline with JSON value successfully parsed
+    pub static ref JSON_ENCODE_LOG_PARSED: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
```
We're counting these things but you haven't added any code to publish them in the internal source. Check out `src/source/internal.rs`

dparton commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,139 @@
+                        if let Value::Object(obj) = json!({"message": log.value.clone()}) {
+                            obj
+                        } else {
+                            panic!()
```
Should this be `unreachable!()`?

ekimekim commented:
PTAL

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,133 @@
+    ) -> Result<(), filter::FilterError> {
+        match event {
+            metric::Event::Telemetry(_) => {
+                // Drop telemetry
```
Rather than drop we should probably forward Telemetry through.

ekimekim replied:
fair enough. wasn't sure of the convention for filters that don't apply to a particular type.

blt replied:
We don't have a concrete policy around this and we should. I lean toward "pass forward unless otherwise documented".

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,133 @@
+}
+
+/// Merge JSON objects, with values from earlier objects in the list overriding later ones.
+/// Note this is not a recursive merge - if the same key is in many objects, we simply take
```
This type of property is ripe for a quickcheck test.

blt commented on src/filter/mod.rs:
```diff
@@ -13,9 +13,11 @@ use util;
 
 pub use self::delay_filter::{DelayFilter, DelayFilterConfig};
 pub use self::flush_boundary_filter::{FlushBoundaryFilter, FlushBoundaryFilterConfig};
+pub use self::json_encode_filter::{JSONEncodeFilter, JSONEncodeFilterConfig};
```
These are now exposed but you'll need to add them to `src/config.rs` to hook up configuration.

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,133 @@
+                        }
+                    );
+                // Combine our various sources of data. Metadata overrides parsed values overrides user-specified.
+                let value = merge_objects(&[&metadata, &log.fields.clone().into(), &value]);
```
I think as a user I'd find this confusing. If my json blob had "time" in it then it'd never come through on the other side. This'd be especially tricky if it _mostly_ did roll through the other side. 

What's stopping us from _adding_ to parsed structures? So, make a `__cernan_metadata` or whatever name field in the parsed blob that contained the contents of metadata? The users' data comes through the other side unchanged, we get our data too. 
Maybe

ekimekim replied:
The primary use-case of this is for searching in logs. In that context the key names need to be short, simple and memorable.
There's probably a good argument for allowing the user to override cernan's metadata rather than vice versa - eg. they probably know the true timestamp of the event better than we do (our view of time is just when we read it).
In fact, yeah I think that ordering makes more sense. The user knows best, anything we can clean from filters (fields) is second best, and our own metadata is third.

blt replied:
> The primary use-case of this is for searching in logs. In that context the key names need to be short, simple and memorable.

That's fair. Once it's merged in master it's not really about the originating use-case anymore and can be applied in surprising ways. That's the perch I'm sitting on when discussing stuff, anyhow. 

> There's probably a good argument for allowing the user to override cernan's metadata rather than vice versa - eg. they probably know the true timestamp of the event better than we do (our view of time is just when we read it).

That's possible to do with a downstream filter. Possible I'm missing some further elaboration you're imagining. 

> In fact, yeah I think that ordering makes more sense. The user knows best, anything we can clean from filters (fields) is second best, and our own metadata is third.

I guess I'm still a little worried about overwriting fields in a payload rather than offsetting them in some way. Adding fields, sure, that's okay but overwriting strikes me as something that could cause surprises, especially for some user that doesn't have logging in mind.

blt commented:
Agreed RE testing. Also, bonus points for adding a benchmark for `JSONEncodeFilter::process`.

ekimekim commented:
PTAL

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,268 @@
+        ).into()
+    }
+
+    quickcheck! {
```
For consistency with the rest of the codebase could you flip this over to use the builder style pattern? A good example of what I mean: https://github.com/postmates/cernan/blob/master/src/buckets.rs#L383

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,268 @@
+            let objs = vecs_to_objs(&vecs);
+            let result = merge_objects(&objs);
+            for (key, result_value) in result {
+                match iter(objs).find(|obj| obj.contains_key(key)) {
```
This `iter(objs)` is going to cause a build error I do believe.

ekimekim replied:
yep, there's a bunch of build issues i'm just dealing with now, i was trying to parallelise fixing those with higher-level feedback :P

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,268 @@
+    // quickcheck and serde_json::map::Map aren't compatible, so we ask quickcheck for
+    // many Vec<(String, String)>s that we turn into maps.
+    fn vecs_to_objs(vecs: &Vec<Vec<(String, String)>>) -> Vec<Map<String, Value>> {
+        iter(vecs).map(|vec|
```
`vecs.iter()`?

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,268 @@
+    // many Vec<(String, String)>s that we turn into maps.
+    fn vecs_to_objs(vecs: &Vec<Vec<(String, String)>>) -> Vec<Map<String, Value>> {
+        iter(vecs).map(|vec|
+            iter(vec).map(|(k, v)| (k.clone(), v.clone().into()))
```
It doesn't matter in-test, but all the cloning here could be avoided if `vecs_to_objs` consumed `vecs`. Right now you're passing by reference and then cloning data from inside the clone.

ekimekim replied:
yeah. I was able to figure that out, but couldn't figure out how to make a consuming iterator such that `k` and `v` would be `String` instead of `&String`.

blt replied:
Try `into_iter`.

blt commented on src/filter/json_encode_filter.rs:
```diff
@@ -0,0 +1,268 @@
+/// Merge JSON objects, with values from earlier objects in the list overriding later ones.
+/// Note this is not a recursive merge - if the same key is in many objects, we simply take
+/// the value from the earliest one.
+fn merge_objects(objs: &[&Map<String, Value>]) -> Map<String, Value> {
```
You're passing `objs` into this function but then the passed values are never re-used. Consider moving `objs` into this function so you can avoid cloning on line 129.

ekimekim commented:
Added tests, including quickchecks for two properties of merge_object (that all keys in the source objects are present in the merged object, and that all keys in the merged object map to the first value found in the list of source objects).
Changed the ordering of the merge so it goes (parsed line, fields, metadata).

PTAL?

ekimekim commented:
rebased onto latest master. no conflicts.

ekimekim commented:
(skipped waiting for tests as there was zero diff between old and new version, just commit history edits)
